src/template/wasm2c.declarations.c

```diff
@@ -67,70 +88,42 @@ static inline bool func_types_eq(const wasm_rt_func_type_t a,
 #define FORCE_READ_FLOAT(var)
 #endif
 
-#if WABT_BIG_ENDIAN
 static inline void load_data(void* dest, const void* src, size_t n) {
   if (!n) {
     return;
   }
-  size_t i = 0;
-  u8* dest_chars = dest;
   wasm_rt_memcpy(dest, src, n);
-  for (i = 0; i < (n >> 1); i++) {
+#if WABT_BIG_ENDIAN
+  u8* dest_chars = dest;
+  for (size_t i = 0; i < (n >> 1); i++) {
     u8 cursor = dest_chars[i];
     dest_chars[i] = dest_chars[n - i - 1];
     dest_chars[n - i - 1] = cursor;
   }
+#endif
 }
-#define LOAD_DATA(m, o, i, s)                   \
-  do {                                          \
-    RANGE_CHECK((&m), m.size - o - s, s);       \
-    load_data(&(m.data[m.size - o - s]), i, s); \
+
+#define LOAD_DATA(m, o, i, s)          \
+  do {                                 \
+    RANGE_CHECK((&m), o, s);           \
```

Comment:
I think this changes the exact range check being performed in big endian mode. While I haven't investigated thoroughly, I suspect this is probably the right check vs. what was already in place (or maybe both checks do the same thing?) 

@keithw @sbc100 @kripken Any chance y'all could have a look here and see if this makes sense to you?

Reply:
if `RANGE_CHECK` had an `o < 0` check, they would be equivalent. we believe this was changed at some point. indeed, the bug seems to trace back to https://github.com/WebAssembly/wabt/commit/407c5f65d172bba3d8b62579468d9c241556ef4b

abstracting out `LEADDR` should make these regressions less likely.

Reply:
The change lgtm given that it makes the tests pass on a BE architecture. My earlier effort to match the behavior was definitely best-effort and untested (in my defense, before that PR, there was *no* range check on active data loads at all...).